benches/src/bin/collect.rs

```diff
@@ -596,7 +596,127 @@ fn slope(x_y: Vec<(u64, u64)>) -> f64 {
+        /// and the last points are almost the same(The difference is < 0.1).
+        Linear,
+        /// When the delta of the last point is much lower than
+        /// the first point, it is a logarithmic chart.
+        Logarithm,
+        /// When the delta of the last point is much more than
+        /// the first point, it is an exponential chart.
+        Exp,
+    }
+
+    #[derive(Clone, Copy)]
+    struct Point {
+        /// Number of elements for the opcode.
+        x: u64,
+        /// Time in `noop`s required to process `x` elements.
+        ///
+        /// Note: If the time to process `noop` opcode is `20` nanoseconds
+        /// and the time required to process `x` elements is `140`,
+        /// the `y` is 7 in this case.
+        y: u64,
+    }
+
+    impl Point {
+        /// The price in `noop` time to process a single element.
+        fn price(&self) -> f64 {
+            (self.y as f64) / (self.x as f64)
+        }
+
+        /// The amount that this operation costs per increase in one `noop` time.
+        fn amount(&self) -> f64 {
+            (self.x as f64) / (self.y as f64)
+        }
+    }
+
+    let linear_regression = linear_regression(x_y.clone());
+
+    let x_y = x_y
+        .into_iter()
+        .map(|(x, y)| Point { x, y })
+        .collect::<Vec<_>>();
+
+    let first = x_y.first().unwrap();
+    let last = x_y.last().unwrap();
+
+    let near_linear = linear_regression * NEAR_LINEAR;
+    let expected_type = if (linear_regression - first.amount()).abs() < near_linear
+        && (linear_regression - last.amount()).abs() < near_linear
+    {
+        Type::Linear
+    } else if first.price() > last.price() {
+        Type::Logarithm
+    } else {
+        Type::Exp
+    };
+
+    match expected_type {
+        Type::Linear => {
+            // The time of the first point is a base.
+            // The minimal charge per element is the worse scenario
+            // and we use it as an amount per unit.
+            let mut x_y = x_y.into_iter();
+            let base = x_y.next().unwrap().y;
+
+            let amount = x_y
+                .map(|p| p.amount())
+                .min_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap();
+            (base, amount as u64)
+        }
+        Type::Logarithm => {
+            // The logarithm function slows down fast, and the point where it becomes more
+            // linear is the base point. After this point we use linear strategy.
+            let mut x_y = x_y.into_iter();
+
+            let mut base = x_y.next().unwrap();
+
+            let amount = x_y
+                .skip_while(|p| {
+                    if p.amount() < linear_regression - base.amount() {
+                        base = *p;
+                        true
+                    } else {
+                        false
+                    }
+                })
+                .map(|p| p.amount())
+                .min_by(|a, b| a.partial_cmp(b).unwrap())
```

Comment:
Wouldn't this potentially leave a range of inputs somewhere between the base and the min(p.amount()) where we undercharge for execution time? would it be safer to just take the slope off the base, even if it overcharges?

Reply:
also nit - isn't this a more complicated expression of simply using `.min()` here?

Reply:
Floats don't support `min` and you need to use `partial_cmp` =(

Remember, the lower value of `amount` means we charge the user more because of `args / dep_per_unit`. So we will only overcharge user, not undercharge.